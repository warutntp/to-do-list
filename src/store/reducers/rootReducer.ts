import { TodoItemModels } from "../../types/TodoItemModels";

export interface RootState {
  todos: TodoItemModels.TodoItemModel[];
}
