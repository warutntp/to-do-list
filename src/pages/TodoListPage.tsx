import React, { useState, useMemo } from "react";
import TodoForm from "../components/TodoForm";
import TodoFilter from "../components/TodoFilter";
import { useSelector } from "react-redux";
import { RootState } from "../store/reducers/rootReducer";
import TodoList from "../components/TodoList";
import { FilterTodoType } from "../types/Enums";

const TodoListPage: React.FC = () => {
  const todos = useSelector((state: RootState) => state.todos);
  const [filter, setFilter] = useState<FilterTodoType>(FilterTodoType.All);

  const filteredTodos = useMemo(() => {
    return todos.filter((todo) => {
      return filter === FilterTodoType.Completed
        ? todo.completed
        : filter === FilterTodoType.Incomplete
        ? !todo.completed
        : true;
    });
  }, [todos, filter]);

  return (
    <div className="container mt-4">
      <div className="d-flex justify-content-center mb-4">
        <h3>To-do List App</h3>
      </div>
      <div className="d-flex justify-content-center todo-list-form-container mb-4">
        <div className="todo-form-container">
          <TodoForm />
        </div>
      </div>
      <div className="d-flex justify-content-end align-items-center mb-4">
        <TodoFilter currentFilter={filter} setFilter={setFilter} />
      </div>
      <hr />
      <TodoList filteredTodos={filteredTodos} />
    </div>
  );
};

export default TodoListPage;
