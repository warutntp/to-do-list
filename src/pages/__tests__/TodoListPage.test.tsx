import React from "react";
import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import TodoListPage from "../TodoListPage";
import { RootState } from "../../store/reducers/rootReducer";

const mockStore = configureStore([]);

const initialState: RootState = {
  todos: [
    {
      slug: "1",
      title: "Test Todo",
      description: "Test description",
      completed: false,
    },
  ],
};

describe("TodoListPage", () => {
  let store: ReturnType<typeof mockStore>;

  beforeEach(() => {
    store = mockStore(initialState);
  });

  it("should render TodoListPage correctly", () => {
    render(
      <Provider store={store}>
        <TodoListPage />
      </Provider>
    );

    expect(screen.getByText("To-do List App")).toBeInTheDocument();
    expect(screen.getByText("Description (Optional)")).toBeInTheDocument();
    expect(
      screen.getByRole("button", { name: /Add Task/i })
    ).toBeInTheDocument();
    expect(screen.getByText("Test Todo")).toBeInTheDocument();
  });
});
