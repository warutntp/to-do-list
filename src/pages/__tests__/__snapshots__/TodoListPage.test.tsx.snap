// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoListPage should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="container mt-4"
  >
    <div
      class="d-flex justify-content-center mb-4"
    >
      <h3>
        To-do List App
      </h3>
    </div>
    <div
      class="d-flex justify-content-center bg-primary mb-4 p-5 rounded"
    >
      <div
        class="todo-form-container bg-white p-3 rounded col-12 col-md-6"
      >
        <form
          class="mb-3"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="title"
            >
              Title
            </label>
            <input
              class="form-control placeholder-italic"
              id="title"
              placeholder="Please enter todo title"
              required=""
              type="text"
              value=""
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="description"
            >
              Description (Optional)
            </label>
            <textarea
              class="form-control placeholder-italic"
              id="description"
              placeholder="Enter todo description"
              rows="3"
            />
          </div>
          <div
            class="mb-3"
          >
            <button
              class="btn btn-primary w-100"
              type="submit"
            >
              Add Task
            </button>
          </div>
        </form>
      </div>
    </div>
    <div
      class="d-flex justify-content-end align-items-center mb-4"
    >
      <div
        class="btn-group"
      >
        <button
          class="btn btn-primary"
          type="button"
        >
          All
        </button>
        <button
          class="btn btn-outline-primary"
          type="button"
        >
          Incomplete
        </button>
        <button
          class="btn btn-outline-primary"
          type="button"
        >
          Completed
        </button>
      </div>
    </div>
    <hr />
    <div
      class="table-responsive"
    >
      <table
        class="table table-hover table-striped"
      >
        <thead>
          <tr>
            <th
              class="col-1 text-center"
              scope="col"
            >
               # 
            </th>
            <th
              class="col-3 text-wrap-break"
              scope="col"
            >
              Title
            </th>
            <th
              class="col-4 text-wrap-break"
              scope="col"
            >
              Description
            </th>
            <th
              class="col-2 text-center"
              scope="col"
            >
              Completed
            </th>
            <th
              class="col-2 text-center"
              scope="col"
            >
              Action
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              class="col-1 text-center"
            >
              1
            </td>
            <td
              class="col-3 text-wrap-break"
            >
              Test Todo
            </td>
            <td
              class="col-4 text-wrap-break"
            >
              Test description
            </td>
            <td
              class="col-2"
            >
              <div
                class="form-check form-switch d-flex justify-content-center"
              >
                <input
                  class="form-check-input"
                  id="1"
                  type="checkbox"
                />
              </div>
            </td>
            <td
              class="text-center"
            >
              <button
                class="btn py-0"
                title="Delete"
                type="button"
              >
                <i
                  class="bi bi-trash bi-danger"
                />
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
